src/page/page_cache.rs

```diff
@@ -359,7 +395,211 @@ impl<M> PageCache<LockFreeLog, M>
+                    Update::Del => {
+                        snapshot.pt.remove(&prepend.pid);
+                        snapshot.free.push(prepend.pid);
+                    }
+                    Update::Alloc => {
+                        snapshot.free.retain(|&pid| pid != prepend.pid);
+                        if prepend.pid > snapshot.max_pid {
+                            snapshot.max_pid = prepend.pid;
+                        }
+                    }
+                }
+            }
+        }
+        snapshot.free.sort();
+        snapshot.free.reverse();
+        snapshot.max_lid = current_stable;
+        snapshot.recovery = recovery;
+
+        let raw_bytes = serialize(&snapshot, Infinite).unwrap();
+        let bytes = if self.config().get_use_compression() {
+            compress(&*raw_bytes, 5).unwrap()
+        } else {
+            raw_bytes
+        };
+        let crc64: [u8; 8] = unsafe { std::mem::transmute(crc64::crc64(&*bytes)) };
+
+        self.last_snapshot.swap(snapshot, SeqCst);
+
+        let prefix = self.snapshot_prefix();
+        let path_1 = format!("{}.{}.in___motion", prefix, current_stable);
+        let path_2 = format!("{}.{}", prefix, current_stable);
+        let mut f = std::fs::OpenOptions::new()
+            .write(true)
+            .create(true)
+            .open(&path_1)
+            .unwrap();
+
+        // write the snapshot bytes, followed by a crc64 checksum at the end
+        f.write_all(&*bytes).unwrap();
+        f.write_all(&crc64).unwrap();
+        f.sync_all().unwrap();
+        drop(f);
+
+        if let Err(e) = std::fs::rename(path_1, &path_2) {
+            error!("failed to write snapshot: {}", e);
+        }
+
+        // clean up any old snapshots
+        let prefix_glob = format!("{}.[0-9]*", prefix);
+        for entry in glob::glob(&*prefix_glob).unwrap() {
+            if let Ok(entry) = entry {
+                if entry.to_str().unwrap() != &*path_2 {
+                    info!("removing old snapshot file {:?}", entry);
+                    std::fs::remove_file(entry).unwrap();
+                }
+            }
+        }
+    }
+
+    fn snapshot_prefix(&self) -> String {
+        let config = self.config();
+        config.get_snapshot_path().or(config.get_path()).unwrap_or(
+            "rsdb".to_owned(),
+        )
+    }
+
+    fn read_snapshot(&mut self) -> Snapshot<R> {
+        let prefix = self.snapshot_prefix();
+        let prefix_glob = format!("{}*snapshot", prefix);
+        let mut candidates = vec![];
+        for entry_res in glob::glob(&*prefix_glob).unwrap() {
+            if let Ok(entry_pb) = entry_res {
+                if !entry_pb.to_str().unwrap().contains(".in___motion") {
+                    candidates.push(entry_pb);
+                }
+            }
+        }
+        if candidates.is_empty() {
+            return Snapshot::default();
+        }
+
+        candidates.sort_by_key(|path| path.metadata().unwrap().created().unwrap());
```

Comment:
again, we are sort of buggy here, because we're just sorting the files by created time and picking the newest one, but it's probably OK, and doesn't impact overall system correctness, since this is all an optimization anyway.